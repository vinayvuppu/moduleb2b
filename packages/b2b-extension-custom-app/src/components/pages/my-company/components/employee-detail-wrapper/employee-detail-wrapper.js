import PropTypes from 'prop-types';
import React from 'react';
import { LoadingSpinner, Spacings } from '@commercetools-frontend/ui-kit';

import EmployeeDetailConnector from '../employee-detail-connector';
import CompanyDetailsConnector from '../../../companies/components/company-details-connector';
import UserNotAssociatedCompany from '../../../../common/UserNotAssociatedCompany';
import B2BApolloClientContext from '../../../../common/b2b-apollo-client-context';

export class EmployeeDetailWrapper extends React.PureComponent {
  static displayName = 'EmployeeDetailWrapper';

  static propTypes = {
    projectKey: PropTypes.string.isRequired,
    children: PropTypes.func.isRequired,
  };

  render() {
    return (
      <B2BApolloClientContext.Consumer>
        {({ apolloClient }) => {
          return (
            <EmployeeDetailConnector
              projectKey={this.props.projectKey}
              apolloClient={apolloClient}
            >
              {({ employeeFetcher }) => {
                if (employeeFetcher.isLoading) {
                  return (
                    <Spacings.Stack scale="m" alignItems="center">
                      <LoadingSpinner />
                    </Spacings.Stack>
                  );
                }
                if (
                  !employeeFetcher.employee ||
                  !employeeFetcher.employee.customerGroup
                ) {
                  return <UserNotAssociatedCompany />;
                }
                const { customerGroup, ...rest } = employeeFetcher.employee;

                return (
                  <CompanyDetailsConnector
                    projectKey={this.props.projectKey}
                    companyId={customerGroup.key}
                  >
                    {({ companyFetcher }) => {
                      if (companyFetcher.isLoading) {
                        return (
                          <Spacings.Stack scale="m" alignItems="center">
                            <LoadingSpinner />
                          </Spacings.Stack>
                        );
                      }
                      return this.props.children({
                        company: companyFetcher.company,
                        employee: rest,
                      });
                    }}
                  </CompanyDetailsConnector>
                );
              }}
            </EmployeeDetailConnector>
          );
        }}
      </B2BApolloClientContext.Consumer>
    );
  }
}

export default EmployeeDetailWrapper;
