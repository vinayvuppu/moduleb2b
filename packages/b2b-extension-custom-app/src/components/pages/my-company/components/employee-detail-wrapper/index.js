export { default } from './employee-detail-wrapper';
