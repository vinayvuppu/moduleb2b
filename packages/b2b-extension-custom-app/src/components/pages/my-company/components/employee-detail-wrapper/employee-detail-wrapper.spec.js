import React from 'react';
import { shallow } from 'enzyme';
import { LoadingSpinner } from '@commercetools-frontend/ui-kit';
import { EmployeeDetailWrapper } from './employee-detail-wrapper';
import B2BApolloClientContext from '../../../../common/b2b-apollo-client-context';

const ChildComponent = () => <div>{'ChildComponent'}</div>;
ChildComponent.displayName = 'ChildComponent';
const createB2BApolloClientContextProps = () => ({ apolloClient: {} });

const createTestProps = customProps => ({
  projectKey: 'projectkey1',
  children: jest.fn(() => <ChildComponent />),
  ...customProps,
});
const createEmployee = () => ({
  id: 'employee-id1',
  customerGroup: {
    id: 'customer-group-id-1',
    key: 'company-id-1',
    name: 'Company name',
  },
  firstName: 'employee name',
  lastName: 'employee last name',
  email: 'foo@bar.com',
});

const createCompany = () => ({
  id: 'company-id-1',
});

const createEmployeeFetcher = custom => ({
  isLoading: false,
  employee: createEmployee(),
  ...custom,
});

const createCompanyFetcher = custom => ({
  isLoading: false,
  company: createCompany(),
  ...custom,
});

const createListConnectorProps = customProps => ({
  employeeFetcher: createEmployeeFetcher(),
  ...customProps,
});

const createCompanyDetailsConnectorProps = customProps => ({
  companyFetcher: createCompanyFetcher(),
  ...customProps,
});

describe('rendering', () => {
  let props;
  let wrapper;

  describe('when loading employee', () => {
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<EmployeeDetailWrapper {...props} />)
        .find(B2BApolloClientContext.Consumer)
        .renderProp('children')(createB2BApolloClientContextProps())
        .renderProp('children')(
        createListConnectorProps({
          employeeFetcher: createEmployeeFetcher({ isLoading: true }),
        })
      );
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render `<LoadingSpinner>', () => {
      expect(wrapper).toRender(LoadingSpinner);
    });
  });

  describe('when loading company', () => {
    beforeEach(() => {
      props = createTestProps();
      wrapper = shallow(<EmployeeDetailWrapper {...props} />)
        .find(B2BApolloClientContext.Consumer)
        .renderProp('children')(createB2BApolloClientContextProps())
        .renderProp('children')(createListConnectorProps())
        .renderProp('children')(
        createCompanyDetailsConnectorProps({
          companyFetcher: { isLoading: true },
        })
      );
    });

    it('should match snapshot', () => {
      expect(wrapper).toMatchSnapshot();
    });

    it('should render `<LoadingSpinner>', () => {
      expect(wrapper).toRender(LoadingSpinner);
    });
  });
  describe('when loaded', () => {
    describe('when the user does not exist as employee', () => {
      beforeEach(() => {
        props = createTestProps();
        wrapper = shallow(<EmployeeDetailWrapper {...props} />)
          .find(B2BApolloClientContext.Consumer)
          .renderProp('children')(createB2BApolloClientContextProps())
          .renderProp('children')(
          createListConnectorProps({
            employeeFetcher: createEmployeeFetcher({
              isLoading: false,
              employee: undefined,
            }),
          })
        );
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should not render children', () => {
        expect(props.children).not.toHaveBeenCalled();
      });
    });
    describe('when the user does not have company associated', () => {
      beforeEach(() => {
        props = createTestProps();
        wrapper = shallow(<EmployeeDetailWrapper {...props} />)
          .find(B2BApolloClientContext.Consumer)
          .renderProp('children')(createB2BApolloClientContextProps())
          .renderProp('children')(
          createListConnectorProps({
            employeeFetcher: createEmployeeFetcher({
              isLoading: false,
              employee: {
                id: 'abc',
                company: undefined,
              },
            }),
          })
        );
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should not render children', () => {
        expect(props.children).not.toHaveBeenCalled();
      });
    });
    describe('when success', () => {
      beforeEach(() => {
        props = createTestProps();
        wrapper = shallow(<EmployeeDetailWrapper {...props} />)
          .find(B2BApolloClientContext.Consumer)
          .renderProp('children')(createB2BApolloClientContextProps())
          .renderProp('children')(createListConnectorProps())
          .renderProp('children')(createCompanyDetailsConnectorProps());
      });

      it('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
      });

      it('should render children', () => {
        expect(props.children).toHaveBeenCalledWith({
          company: { id: 'company-id-1' },
          employee: {
            id: 'employee-id1',
            firstName: 'employee name',
            lastName: 'employee last name',
            email: 'foo@bar.com',
          },
        });
      });
    });
  });
});
