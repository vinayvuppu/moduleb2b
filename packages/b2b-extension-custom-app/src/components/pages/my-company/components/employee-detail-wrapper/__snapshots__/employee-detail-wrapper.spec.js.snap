// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`rendering when loaded when success should match snapshot 1`] = `<ChildComponent />`;

exports[`rendering when loaded when the user does not exist as employee should match snapshot 1`] = `<UserNotAssociatedCompany />`;

exports[`rendering when loaded when the user does not have company associated should match snapshot 1`] = `<UserNotAssociatedCompany />`;

exports[`rendering when loading company should match snapshot 1`] = `
<Stack
  alignItems="center"
  scale="m"
>
  <LoadingSpinner
    scale="l"
  />
</Stack>
`;

exports[`rendering when loading employee should match snapshot 1`] = `
<Stack
  alignItems="center"
  scale="m"
>
  <LoadingSpinner
    scale="l"
  />
</Stack>
`;
